### PR TITLE
[FLINK-32309][sql-gateway] Use independent resource manager for table environment

### PR DESCRIPTION
**Pull request by FangYongs**
## What is the purpose of the change

This PR aims to create new resource manager for table environment in sql gateway

## Brief change log

  - Create new resource manager which will share the classloader and local path with the previous manager
  - Create table environment for each operation in gateway with the new resource manager
  - Add jars and functions related operation in gateway which will be performed on the session resource manager.

## Verifying this change

This change added tests and can be verified as follows:

  - Updated `function.q` to test that previous job will not affect the jars in gateway

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by FangYongs** [comment]
Hi @fsk119 Please help to review this PR when you are free, thanks

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/OperationExecutor.java [hunk 1]
```diff
@@ -324,15 +332,16 @@ public TableEnvironmentInternal getTableEnvironment() {
 
         final Executor executor =
                 lookupExecutor(streamExecEnv, sessionContext.getUserClassloader());
+        ResourceManager resourceManager = sessionContext.getSessionState().resourceManager.copy();
```
Comment (fsk119) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/OperationExecutor.java:335]: Currently the ResourceManager uses lazy loading strategy to load the used resources. That is, the resource manager will add the function jar resources(*the function is not temporary*) when the query uses this. If we use a seperate resource manager here, it has some side efforts:

1. we will keep downloading from the remote function jar when using it
2. it's hard to track whether the resource manager contains the resource with the same resource uri.

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. Could you share the motivation about the PR? I don't understand why we need to use independent rm here?

**Comment by FangYongs** [comment]
> 

I would like to use independent RM for each query in sql-gateway to avoid submitting job with unnecessary jars. For example, we create a function as follows:
`create function aaa AS 'xxx' using jar 'xxx/a.jar'`
Users can submit a query with aaa function as follows
`select x1, aaa(x2), x3 from table ...;`
The jar file `xxx/a.jar` will be added to RM for the query.

If we use the shared RM in session for sql-gateway, when user submit a query without function `aaa` as follows
`select x1, x2, x3 from table ...;`
This job will always has the `xxx/a.jar` in its path, and job manager will create user classloader with the `xxx/a.jar` for the new job even it does not need it.

**Comment by FangYongs** [comment]
@fsk119 As we discussed, I have create another PR [22938](https://github.com/apache/flink/pull/22938) for temporary function, please help to review it, thanks

**Comment by KarmaGYZ** [comment]
@FangYongs Would you like to update this PR as #22938 has been merged?

**Comment by FangYongs** [comment]
Thanks @KarmaGYZ , I will rebase master and work on this issue

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/OperationExecutor.java [hunk 2]
```diff
@@ -176,14 +184,17 @@ public ResultFetcher configureSession(OperationHandle handle, String statement)
             return callSetOperation(tableEnv, handle, (SetOperation) op);
         } else if (op instanceof ResetOperation) {
             return callResetOperation(handle, (ResetOperation) op);
+        } else if (op instanceof AddJarOperation) {
+            return callExecutableOperation(handle, (ExecutableOperation) op);
```
[the remark this answers is not in the document]
Comment (FangYongs) [reply flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/OperationExecutor.java:188]: You're right, we don't need it, I'll remove this

**Review by KarmaGYZ — APPROVED** [review]

**Pull request merged**